Dockerfile

```diff
@@ -110,21 +110,47 @@ RUN file -b   /bin/invoicer
 RUN du        /bin/invoicer
 
 
+# This stage is used to generate /etc/{group,passwd,shadow} files & avoid RUN-ing commands in the `final` layer,
+#   which would break cross-compiled images.
+FROM alpine:3.11 AS perms
+
+ARG USER=invoicer
+ARG DIR=/data/
+
+# NOTE: Default GID == UID == 1000
+RUN adduser --disabled-password \
+            --home ${DIR} \
+            --gecos "" \
+            ${USER}
+
+
 # This is a final stage, destined to be distributed, if successful
 FROM alpine:3.11 AS final
 
+ARG USER=invoicer
+ARG DIR=/data/
```

Comment:
You don't need to duplicate the ARG values if you hoist them outside the `FROM` block.

e.g:

```Dockerfile
ARG USER=invoicer
ARG DIR=/data/

FROM alpine:3.11 AS perms

ARG USER
ARG DIR

FROM alpine:3.11 AS final

ARG USER
ARG DIR
```

https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

Reply:
Beautiful, I've had no idea that's possible! 🤩